dask/array/core.py

```diff
@@ -1876,6 +1876,10 @@ def __setitem__(self, key, value):
         if value is np.ma.masked:
             value = np.ma.masked_all((), dtype=self.dtype)
 
+        if not is_dask_collection(value) and np.isnan(value).any():
+            if issubclass(self.dtype.type, Integral):
+                raise ValueError("cannot convert float NaN to integer")
```

Comment:
I might expect this to raise a `TypeError`, per the python [docs](https://docs.python.org/3/library/exceptions.html#TypeError).

Reply:
I went with `ValueError` purely motivated by parity with NumPy, but after reading the `TypeError` docs I'd agree with you that it's better. Do you have an opinion on what should be prioritized?

Reply:
Ah, good point, I had missed that numpy used `ValueError`. In that case, consider my nit retracted!